src/lavinmq/mqtt/retain_store.cr

```diff
@@ -0,0 +1,144 @@
+          msg_file_name = make_file_name(topic)
+          unless msg_file_segments.delete(msg_file_name)
+            Log.warn { "msg file for topic #{topic} missing, dropping from index" }
+            next
+          end
+          index.insert(topic, msg_file_name)
+          Log.debug { "restored #{topic}" }
+          msg_count += 1
+        end
+
+        # TODO: Device what's the truth: index file or msgs file. Mybe drop the index file and rebuild
+        # index from msg files?
+        unless msg_file_segments.empty?
+          Log.warn { "unreferenced messages: #{msg_file_segments.join(",")}" }
+        end
+        # TODO: delete unreferenced messages?
+        Log.info { "restoring index done, msg_count = #{msg_count}" }
+      end
+
+      def retain(topic : String, body_io : ::IO, size : UInt64) : Nil
+        @lock.synchronize do
+          Log.debug { "retain topic=#{topic} body.bytesize=#{size}" }
+          # An empty message with retain flag means clear the topic from retained messages
+          if size.zero?
+            delete_from_index(topic)
+            return
+          end
+
+          unless msg_file_name = @index[topic]?
+            msg_file_name = make_file_name(topic)
+            add_to_index(topic, msg_file_name)
+          end
+
+          File.open(File.join(@dir, msg_file_name), "w+") do |f|
+            f.sync = true
+            ::IO.copy(body_io, f)
+          end
+        end
+      end
+
+      private def write_index
+        tmp_file = File.join(@dir, "#{INDEX_FILE_NAME}.next")
+        File.open(tmp_file, "w+") do |f|
+          @index.each do |topic, _filename|
+            f.puts topic
+          end
+        end
+        File.rename tmp_file, File.join(@dir, INDEX_FILE_NAME)
+      ensure
+        FileUtils.rm_rf tmp_file unless tmp_file.nil?
+      end
+
+      private def add_to_index(topic : String, file_name : String) : Nil
+        @index.insert topic, file_name
+        @index_file.puts topic
+        @index_file.flush
+      end
+
+      private def delete_from_index(topic : String) : Nil
+        if file_name = @index.delete topic
+          Log.trace { "deleted '#{topic}' from index, deleting file #{file_name}" }
+          File.delete? File.join(@dir, file_name)
+        end
+      end
+
+      def each(subscription : String, &block : String, Bytes -> Nil) : Nil
+        @lock.synchronize do
+          @index.each(subscription) do |topic, file_name|
+            block.call(topic, read(file_name))
+          end
+        end
+        nil
+      end
+
+      private def read(file_name : String) : Bytes
+        # @lock.synchronize do
+        File.open(File.join(@dir, file_name), "r") do |f|
+          body = Bytes.new(f.size)
+          f.read_fully(body)
+          body
+        end
+        # end
+      end
+
+      def retained_messages
+        @lock.synchronize do
+          @index.size
+        end
+      end
+
+      @hasher = Digest::MD5.new
```

Comment:
Md5 is fine? We normally use sha1

Reply:
we can use sha1 if its a security thing, MD5 is faster and gives shorter hashes.. and i get the security aspect, but for filenames maybe it can be sufficient? but maybe it is better to be consistent.